src/server/coprocessor/executor/topn.rs

```diff
@@ -0,0 +1,325 @@
+        if expr.get_tp() == ExprType::ColumnRef {
+            self.col_ids.insert(box_try!(expr.get_val().decode_i64()));
+        } else {
+            for sub_expr in expr.get_children() {
+                try!(self.visit(sub_expr));
+            }
+        }
+        Ok(())
+    }
+}
+
+pub struct TopNExecutor<'a> {
+    order_by: Rc<Vec<ByItem>>,
+    columns: Vec<ColumnInfo>,
+    heap: Option<TopNHeap>,
+    executed: bool,
+    res: Option<Vec<SortRow>>,
+    cursor: usize,
+
+    src: Box<Executor + 'a>,
+    ctx: Rc<EvalContext>,
+}
+
+impl<'a> TopNExecutor<'a> {
+    pub fn new(mut meta: TopN,
+               ctx: Rc<EvalContext>,
+               columns_info: &[ColumnInfo],
+               src: Box<Executor + 'a>)
+               -> Result<TopNExecutor<'a>> {
+        let order_by = meta.take_order_by().into_vec();
+
+        let mut visitor = ExprColumnRefVisitor::new();
+        for order_by in (&order_by).iter() {
+            try!(visitor.visit(order_by.get_expr()));
+        }
+        let columns = columns_info.iter()
+            .filter(|col| visitor.col_ids.get(&col.get_column_id()).is_some())
+            .cloned()
+            .collect();
+
+        Ok(TopNExecutor {
+            order_by: Rc::new(order_by),
+            heap: Some(try!(TopNHeap::new(meta.get_limit() as usize))),
+            columns: columns,
+            executed: false,
+            res: None,
+            cursor: 0,
+            ctx: ctx,
+            src: src,
+        })
+    }
+
+    fn inner_next(&mut self) -> Result<Option<()>> {
+        if let Some(row) = try!(self.src.next()) {
+            let mut eval = Evaluator::default();
+            try!(inflate_with_col(&mut eval, &self.ctx, &row.data, &self.columns, row.handle));
+            let mut ob_values = Vec::with_capacity(self.order_by.len());
+            for by_item in self.order_by.as_ref().iter() {
+                let v = box_try!(eval.eval(&self.ctx, by_item.get_expr()));
+                ob_values.push(v);
+            }
+            try!(self.heap.as_mut().unwrap().try_add_row(row.handle,
+                                                         row.data,
+                                                         ob_values,
+                                                         self.order_by.clone(),
+                                                         self.ctx.clone()));
+            Ok(Some(()))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl<'a> Executor for TopNExecutor<'a> {
+    fn next(&mut self) -> Result<Option<Row>> {
+        if !self.executed {
+            while try!(self.inner_next()).is_some() {}
+            self.executed = true;
+            self.res = Some(try!(self.heap.take().unwrap().into_sorted_vec()));
+            self.res.as_mut().unwrap().reverse();
+        }
+        let res = self.res.as_mut().unwrap();
+        if res.is_empty() {
+            return Ok(None);
+        }
+        if let Some(sort_row) = res.pop() {
```

Comment:
`res` is at least one value here. So we can remove line123-124 or pops value directly.